src/providers/ipa/ipa_hbac_common.c

```diff
@@ -533,56 +536,108 @@ hbac_eval_user_element(TALLOC_CTX *mem_ctx,
-     */
+    ipa_domain = get_domains_head(domain);
+    if (ipa_domain == NULL) {
+        ret = EINVAL;
+        goto done;
+    }
+
+    ipa_groups_basedn = ldb_dn_new_fmt(tmp_ctx, sysdb_ctx_get_ldb(domain->sysdb),
+                                        SYSDB_TMPL_GROUP_BASE, ipa_domain->name);
+    if (ipa_groups_basedn == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
     ret = sysdb_search_user_by_name(tmp_ctx, domain, username,
-                                    attrs, &msg);
+                                    NULL, &msg);
     if (ret != EOK) {
         DEBUG(SSSDBG_CRIT_FAILURE,
               "Could not determine user memberships for [%s]\n",
                   users->name);
         goto done;
     }
 
-    el = ldb_msg_find_element(msg, SYSDB_ORIG_MEMBEROF);
-    if (el == NULL || el->num_values == 0) {
+    /*
+     * Get the name attribute of all groups pointed to by the memberof
+     * attribute. This includes both POSIX and non-POSIX groups.
+     */
+    ret = sysdb_asq_search(tmp_ctx, domain, msg->dn,
+                           "("SYSDB_OBJECTCLASS"="SYSDB_GROUP_CLASS")",
+                           SYSDB_MEMBEROF,
+                           attrs,
+                           &m_count, &members);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "sysdb_asq_search failed [%d]: %s\n", ret, sss_strerror(ret));
+        goto done;
+    }
+
+    if (m_count == 0) {
         DEBUG(SSSDBG_TRACE_LIBS, "No groups for [%s]\n", users->name);
         ret = create_empty_grouplist(users);
         goto done;
     }
-    DEBUG(SSSDBG_TRACE_LIBS,
-          "[%d] groups for [%s]\n", el->num_values, users->name);
+    DEBUG(SSSDBG_TRACE_LIBS, "[%zu] groups for [%s]\n", m_count, username);
 
-    users->groups = talloc_array(users, const char *, el->num_values + 1);
+    users->groups = talloc_array(users, const char *, m_count + 1);
     if (users->groups == NULL) {
         ret = ENOMEM;
         goto done;
     }
 
-    for (i = 0; i < el->num_values; i++) {
-        member_dn = (const char *)el->values[i].data;
+    for (size_t i = 0; i < m_count; i++) {
+        fqgroupname = ldb_msg_find_attr_as_string(members[i], SYSDB_NAME, NULL);
+        if (fqgroupname == NULL) {
+            DEBUG(SSSDBG_MINOR_FAILURE,
+                  "Skipping malformed entry [%s]\n",
+                  ldb_dn_get_linearized(members[i]->dn));
+            continue;
+        }
 
-        ret = get_ipa_groupname(users->groups, domain->sysdb, member_dn,
-                                &users->groups[num_groups]);
-        if (ret != EOK && ret != ERR_UNEXPECTED_ENTRY_TYPE) {
+        /* Only groups from the IPA domain can be referenced from HBAC rules. To
+         * avoid evaluating groups which might even have the same name, but come
+         * from a trusted domain, we first copy the DN to a temporary one..
+         */
+        member_group_container = ldb_dn_copy(tmp_ctx, members[i]->dn);
```

Comment:
Maybe the check can be simplified by calling something like

    if (ldb_dn_get_comp_num(members[i]->dn) != exp_comp || ldb_dn_compare(ipa_groups_basedn, members[i]->dn) != 0) {
        continue;
    }

where

    exp_comp = db_dn_get_comp_num(ipa_groups_basedn) + 1;